File: onejit/kind.cpp
/*
 * onejit - JIT compiler in C++
 *
 * Copyright (C) 2018-2021 Massimiliano Ghilardi
 *
 *     This Source Code Form is subject to the terms of the Mozilla Public
 *     License, v. 2.0. If a copy of the MPL was not distributed with this
 *     file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * kind.cpp
 *
 *  Created on Jan 08, 2021
 *      Author Massimiliano Ghilardi
 */

#include <onejit/fmt.hpp>
#include <onejit/kind.hpp>

namespace onejit {

static const Chars kstring[] = {
    "bad",     "void",      "bool",              //
    "int8",    "int16",     "int32",   "int64",  //
    "uint8",   "uint16",    "uint32",  "uint64", //
    "float16", "float32",   "float64",           //
    "ptr",     "archflags", "?",                 //
};

static const char kstringsuffix[] = //
    "\1?\0\1v\0\1e\0"
    "\1b\0\1s\0\1i\0\1l\0"
    "\2ub\2us\2ui\2ul"
    "\2hf\1f\0\2lf"
    "\1p\0\003cmp";

static const Bits kbits[] = {
    Bits0,  Bits0,  Bits1,          // Bad, Void, Bool
    Bits8,  Bits16, Bits32, Bits64, // Int*
    Bits8,  Bits16, Bits32, Bits64, // Uint*
    Bits16, Bits32, Bits64,         // Float*
    Bits64, Bits64,                 // Ptr, ArchFlags
};

static const Group kgroup[] = {
    gVoid,  gVoid,  gBool,         // Bad, Void, Bool
    gInt,   gInt,   gInt,   gInt,  // Int*
    gUint,  gUint,  gUint,  gUint, // Uint*
    gFloat, gFloat, gFloat,        // Float*
    gPtr,   gArch,                 // Ptr, ArchFlags
};

const Chars Kind::string() const noexcept {
  enum _ { n = ONEJIT_N_OF(kstring) };
  uint8_t i = val_ & 0xF;
  if (i >= n) {
    i = n - 1; // "?"
  }
  return kstring[i];
}

const Chars Kind::stringsuffix() const noexcept {
  enum _ { n = ONEJIT_N_OF(kstringsuffix) };
  uint8_t i = val_ & 0xF;
  if (i >= n) {
    i = 0;
  }
  const char *str = &kstringsuffix[i * 3];
  return Chars{str + 1, uint8_t(str[0])};
}

Bits Kind::bits() const noexcept {
  uint8_t i = val_ & 0xF;
  if (i >= ONEJIT_N_OF(kbits)) {
    i = 0;
  }
  return kbits[i] * simdn();
}

Group Kind::group() const noexcept {
  uint8_t i = val_ & 0xF;
  if (i >= ONEJIT_N_OF(kgroup)) {
    i = 0;
  }
  return kgroup[i];
}

const Fmt &operator<<(const Fmt &fmt, Kind kind) {
  fmt << kind.string();
  const size_t n = kind.simdn().val();
  if (n != 1) {
    fmt << 'x' << n;
  }
  return fmt;
}

} // namespace onejit

File: go/constant/kind.go
/*
 * Copyright (C) 2021 Massimiliano Ghilardi
 *
 *     This Source Code Form is subject to the terms of the Mozilla Public
 *     License, v. 2.0. If a copy of the MPL was not distributed with this
 *     file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 *
 * kind.go
 *
 *  Created on: Apr 12, 2021
 *      Author: Massimiliano Ghilardi
 */

package constant

import "github.com/cosmos72/onejit/go/token"

// alias for token.Kind
type Kind = token.Kind

const (
	Invalid = Kind(token.Invalid)

	// types for typed constants
	Bool       = Kind(token.Bool)
	Int        = Kind(token.Int)
	Int8       = Kind(token.Int8)
	Int16      = Kind(token.Int16)
	Int32      = Kind(token.Int32)
	Int64      = Kind(token.Int64)
	Uint       = Kind(token.Uint)
	Uint8      = Kind(token.Uint8)
	Uint16     = Kind(token.Uint16)
	Uint32     = Kind(token.Uint32)
	Uint64     = Kind(token.Uint64)
	Uintptr    = Kind(token.Uintptr)
	Float32    = Kind(token.Float32)
	Float64    = Kind(token.Float64)
	Complex64  = Kind(token.Complex64)
	Complex128 = Kind(token.Complex128)
	String     = Kind(token.String)

	// types for untyped constants
	UntypedBool    = Kind(token.UntypedBool)
	UntypedInt     = Kind(token.UntypedInt)
	UntypedRune    = Kind(token.UntypedRune)
	UntypedFloat   = Kind(token.UntypedFloat)
	UntypedComplex = Kind(token.UntypedComplex)
	UntypedString  = Kind(token.UntypedString)
	UntypedNil     = Kind(token.UntypedNil)

	// aliases
	Byte = Kind(token.Byte)
	Rune = Kind(token.Rune)
)
